docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+- [`write-fonts`](#write-fonts)
+    - [tables and records](#write-tables-records)
+    - [fields and `#[compile(..)]`](#table-fields)
+    - [offsets](#write-offsets)
+    - [parsing and `FromTableRef`](#write-parsing)
+    - [validation](#validation)
+    - [compilation and `FontWrite`](#compilation)
+
+## <a id="overview"></a> overview
+
+These two crates can be thought of as siblings, and they both follow the same
+basic high-level design pattern: they contain a set of generated types, mapping
+*as closely as possible* to the types in the [OpenType spec][opentype],
+alongside hand-written code that uses and is used by those types.
+
+The [`read-fonts`][] crate is focused on efficient read access and parsing, and
+the [`write-fonts`][] crate is focused on compilation. The two crates contain a
+parallel `tables` module, with a nearly identical set of type definitions: for
+instance, [both crates][read-name-record] [contain a][write-name-record] `tables::name::NameRecord` type.
+
+We will examine each of these crates separately.
+
+## <a id="read-fonts"></a> `read-fonts`
+
+In the [`read-fonts`][] crate, we make a distinction between *table* objects and
+*record* objects, and we generate different code for each.
+
+The distinction between a *table* and a *record* is blurry, but the
+specification offers two "general criteria":
+
+> - Tables are referenced by offsets. If a table contains an offset to a
+> sub-structure, the offset is normally from the start of that table.
+> - Records occur sequentially within a parent structure, either within a
+> sequence of table fields or within an array of records of a given type. If a
+> record contains an offset to a sub-structure, that structure is logically a
+> subtable of the record’s parent table and the offset is normally from the start
+> of the parent table.
+>
+> ([The OpenType font file][otff])
+
+### <a id="scalars-detour"></a> A brief detour on scalars and `BigEndian<T>`
+
+#### a description of the problem
+
+Before we dive into the specifics of the tables and records in `read-fonts`, I
+want to talk briefly about how we represent and handle the [basic data types](ot-data-types)
+of which records and tables are composed.
+
+In the font file, these values are all represented in [big-endian][endianness]
+byte order. When we access them, we will need to convert them to the native
+endianness of the host platform. We also need to have some set of types which
+exactly match the memory layout (including byte ordering) of the underlying font
+file; this is necessary for us to take advance of zerocopy semantics (see the
+[zerocopy section](#zerocopy) below.)
+
+This leads us to a situation where we require two distinct types for each
+scalar: a native type that we will use in our program logic, and a
+'raw' type that will represent the bytes in the font file (as well as some
+mechanism to convert between them.)
+
+There are various ways we could express this in Rust. The most straightforward
```

Comment:
Are there other options we considered beyond these two?

Reply:
I considered three, total: `u16`/`RawU16`, `U16<BE>`/`U16<LE>`, and `u16`/`BigEndian<u16>`.